### PR TITLE
Improve PMV Calculations to consider cooling credit for elevated air velocity

### LINKED ISSUE #8222: Improving PMV model
Issue overview
--------------
Current PMV models in EnergyPlus do not reflect the latest ASHRAE Standard 55-2017 which provides cooling credit for elevated air velocity. The new PMV is crucial to support passive cooling technologies such as natural ventilation, ceiling fan , and portable fan. 

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by xuanluo113**
Pull request overview
---------------------
In this new feature, we implemented the following adjustments according to ASHRAE Standard 55-2017 in EnergyPlus to facilitate more accurate comfort modeling:

1. Add the cooling effect (CE) calculations, according to ASHRAE Standard 55-2017 NORMATIVE APPENDIX D. 

2.  Add a modified set of PMV/PPD calculations that consider the cooling effect of elevated air speed. Keep the original PMV/PPD calculations.
3. Calculate the percentage of thermally dissatisfied people with the ankle draft at 0.1 m above floor level.

### Code changes
1. Added the above features with IDD changes.
2. Reorganized the thermal comfort calculations to separate the PMV, PPD, SET calculations as stand-alone calls to reuse. New functions are listed in the changes of `ThermalComfort.hh`.
3. Modified the Pierce two-node SET calculations to reflect the changes in ASH55-2017. The changes were related to the autogenerated air speed by physical movement and some unit transformation that was incorrect. 

The new results are verified in the unit test, referring to the ASH55-2017 APPENDIX D - Validation Table for SET Computer Model.

In detail, the differences between the original SET model and the ASH55-2017 Appendix program includes:

1. Differences in the thermal const in the two programs, such as MinSkinBloodFlow
(minimum value of skin blood flow) and RegSweatMax (Max. value of regulatory sweating).
2. The Air Velocity has a lower bound of 0.137 m/s in the original program, but 0.1 m/s in the ASHRAE Appendix program.
3. The original Vapor Pressure was calculated in Pa, while the ASHRAE Appendix program calculates vapor pressure in Torr. Parameters are not consistent due to the unit conversion.
4. Minor differences in algorithms to calculate some intermediate variables such as the body thermal capacities and ratio heat loss sweating to max heat loss sweating.
5. Different iterative solver and tolerance to calculate SET.

### Diffs
1. EIO diffs are due to the changes to the People summary of adding two variables, `Cooling Effect Calculation` and `Ankle Draft Calculation`.
2. The ESO diffs are triggered by the changes of the program to calculate SET.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by nrel-bot-3** [comment]
@xuanluo113 @lgentile it has been 28 days since this pull request was last updated.

**Comment by xuanluo113** [comment]
@Myoldmopar This branch is ready for review, including code changes, tests, and documentation. Would you mind suggesting reviewers to the feature? Thanks.

File: idd/Energy+.idd.in
Comment (mjwitte) [inline idd/Energy+.idd.in]: Either delete this note or change fifth --> sixth
Similar for the next field.

File: idd/Energy+.idd.in
Comment (mjwitte) [inline idd/Energy+.idd.in]: Hmmm, this new field would work better if it was moved up to be A14 before Thermal Comfort Model 1 Type. But that would add a transition requirement. Since the thermal comfort model fields are not extensible (because there's a limited set of options). So, I guess it's fine here.

File: src/EnergyPlus/InternalHeatGains.cc
Comment (mjwitte) [inline src/EnergyPlus/InternalHeatGains.cc]: This looks wrong `cAlphaFieldNames(7) + '=' + AlphaName(8)` should both be 7 or 8, I think.

File: src/EnergyPlus/InternalHeatGains.cc
Comment (mjwitte) [inline src/EnergyPlus/InternalHeatGains.cc]: Yes, this is confusing, but it's ok (it took me a few minutes to see this). Item1 == 1 refers to the first Zone for this People object. 
`for (Item1 = 1; Item1 <= PeopleObjects(Item).NumOfZones; ++Item1) {`
If a given People object references a ZoneList, the fields in the object get processed multiple times, so it makes sense to throw warnings/errors only on the first pass.
I could think of better variable names that `Item1` here, but probably not worth changing at this point.

File: src/EnergyPlus/InternalHeatGains.cc
Comment (mjwitte) [inline src/EnergyPlus/InternalHeatGains.cc]: `ThermalComfortRuns` is a confusing name here, because it's only true for a subset of the thermal comfort models (all of them except AdaptiveASH55 and AdaptiveCEN15251). If I'm following correctly, the 5 in this bool all require some additional input fields. So maybe there's a better variable name to distinguish this from `UsingThermalComfort` higher up?

File: src/EnergyPlus/InternalHeatGains.cc
Comment (mjwitte) [inline src/EnergyPlus/InternalHeatGains.cc]: I don't see the point of `MustInpSch`. I know it's in the old code, but it's true for *every* thermal comfort option, so it seems redundant. Maybe you can see how it's useful.

File: src/EnergyPlus/InternalHeatGains.cc
Comment (mjwitte) [inline src/EnergyPlus/InternalHeatGains.cc]: Only the last group should have `\n` (line break) so it should be removed here and in the KSU block above.

File: src/EnergyPlus/ThermalComfort.cc
Comment (mjwitte) [inline src/EnergyPlus/ThermalComfort.cc:52]: Hmmm, I think this would be the first use of `boost` within the main EnergyPlus code. @Myoldmopar ok with that? I know it's used in Kiva, but I think that's the only place.

**Review by mjwitte — COMMENTED** [review]
@xuanluo113 Some initial comments. I stopped at the top of ThermalComfort.cc

**Comment by xuanluo113** [comment]
@mjwitte Thanks. Comments are addressed.

File: src/EnergyPlus/InternalHeatGains.cc
Comment (mjwitte) [inline src/EnergyPlus/InternalHeatGains.cc:1016]: This error message could be improved. In fact, several of the existing errors could be better. These can be fixed later. See https://github.com/NREL/EnergyPlus/issues/8537
There were other existing problems with `errflag` that I've fixed here and pushed up to prevent errors in People from throwing errors for Lights, Electric Equipment, etc.

File: src/EnergyPlus/ThermalComfort.cc
Comment (mjwitte) [inline src/EnergyPlus/ThermalComfort.cc:798]: These can all be `constexpr` right?

**Review by mjwitte — COMMENTED** [review]
@xuanluo113 I've checked in some doc cleanups and changed the new idf Version to 9.5 (in anticipation of #8532 likely merging before this PR.

There are some follow-up items to fix later so we don't hold this up for I/O Freeze. See #8537.

I plan to do some more testing tomorrow of the actual model outputs before merging this.

File: src/EnergyPlus/ThermalComfort.cc
Comment (mjwitte) [inline src/EnergyPlus/ThermalComfort.cc:1174]: @Myoldmopar Please review.

**Comment by mjwitte** [comment]
Some of the diffs in the Pierce outputs are quite large.

1ZoneUncontrolled_Win_ASH55_Thermal_Comfort
Winter design day, ZONE ONE PEOPLE:Zone Thermal Comfort Pierce Model Standard Effective Temperature [C] changes from -4.58 to -8.95. 
The rest of the simulation the diffs range from -1 to +0.9.

DynamicClothing
WEST ZONE:Zone Thermal Comfort Pierce Model Thermal Sensation Index [], on 10/04 17:00-19:00 changes from 3 to 4.2. 
The rest of the simulation diffs range from -0.3 to +0.8.

TermReheatSurfTC
WEST ZONE:Zone Thermal Comfort Pierce Model Thermal Sensation Index [] diffs range from -0.04 to +0.4.
WEST ZONE:Zone Thermal Comfort Pierce Model Standard Effective Temperature [C], max diffs are +1.3 (values near 30) on 07/21.
The rest of the simulation diffs range from 0.05 to 0.9.

So, for conditioned spaces, the diffs seem reasonable. Can you confirm the large change in SET for the unconditioned winter day?

If possible, at the top where the eso diffs are mentioned, please add a description of when large differences can be expected (more affected by cold or hot, humid or dry)? Something to help users understand the potential impact on results would be helpful.

**Comment by mjwitte** [comment]
@RKStrand Do you want to review this before it goes in?

**Comment by xuanluo113** [comment]
@mjwitte These large diffs are expected. The SET algorithm is complicated with [two-layer converging processes](https://github.com/NREL/EnergyPlus/blob/28fbc1dce17dc40595a3532fc4743de08ca44db0/src/EnergyPlus/ThermalComfort.cc#L860). And the code changes and bug fixings to this algorithm, listed in the descriptions of this PR, are not all minor, and the converging process enlarged this difference.

The results are fully tested based on the table below (ref: ASHRAE 55-2017 Appendix D), and about half of the cases are included in the unit test. Based on my experiments testing the original code, the cases with high or low temperature, and high or low humidity, turn to deviate with the validation table more. But I can't say this indicates only edge cases are affected or more likely to be affected.

Screen Shot 2021-02-16 at 9 34 25 AM

**Comment by RKStrand** [comment]
@mjwitte Sure, except I won't be able to get to that today.  Hopefully tomorrow I'll have a chance to look at this...

File: src/EnergyPlus/InternalHeatGains.cc
Comment (xuanluo113) [reply src/EnergyPlus/InternalHeatGains.cc]: This is fixed.

File: src/EnergyPlus/InternalHeatGains.cc
Comment (xuanluo113) [reply src/EnergyPlus/InternalHeatGains.cc]: This is fixed.

File: src/EnergyPlus/InternalHeatGains.cc
Comment (xuanluo113) [reply src/EnergyPlus/InternalHeatGains.cc]: `MustInpSch` is not used and deleted.

File: src/EnergyPlus/InternalHeatGains.cc
Comment (xuanluo113) [reply src/EnergyPlus/InternalHeatGains.cc]: Fixed.

File: idd/Energy+.idd.in
Comment (xuanluo113) [reply idd/Energy+.idd.in]: Yes. I put it at the bottom as too many transitions change would be triggered to add a field in the middle of `People` object.

File: idd/Energy+.idd.in
Comment (xuanluo113) [reply idd/Energy+.idd.in]: Fixed.

File: src/EnergyPlus/ThermalComfort.cc
Comment (mjwitte) [reply src/EnergyPlus/ThermalComfort.cc:798]: I will add this to the followup issue #8537.

**Review by mjwitte — APPROVED** [review]
@xuanluo113 Looking at outputs for new example file 1ZoneUncontrolled_Win_ASH55_Thermal_Comfort
Most of the time ZONE ONE PEOPLE:Zone Thermal Comfort ASHRAE 55 Ankle Draft PPD [](Hourly) is a similar order of magnitude as ZONE ONE PEOPLE:Zone Thermal Comfort ASHRAE 55 Elevated Air Speed Cooling Effect Adjusted PPD [](Hourly), but during the summer design day (07/21) the fanger and air speed PPDs are 99 or 100, but the ankle draft PPD is less than 1.0. Is that possible? If there's an error here, just add it to #8537.

I'm done reviewing this PR. @Myoldmopar it's all yours.

**Comment by xuanluo113** [comment]
@mjwitte The ankle draft PPD measures the dissatisfaction level of the draft, or the chillness at the ankle level with low temperature, high relative air velocity. It has nothing to do with other sensations than the draft. In (unconditioned) summer, the temperature is high enough to prevent draft risk, so the percent of dissatisfaction is near zero.
I can add more explanation of this in the engineering reference. I'll add it to #8537.

File: idd/Energy+.idd.in
Comment (Myoldmopar) [inline idd/Energy+.idd.in:20898]: OK, so new key choices and then a few new fields on the end.  Doesn't look like any transition needed :+1:

File: src/EnergyPlus/DataHeatBalance.cc
Comment (Myoldmopar) [inline src/EnergyPlus/DataHeatBalance.cc:583]: A few new "any" flags added, OK.

File: src/EnergyPlus/DataHeatBalance.hh
Comment (Myoldmopar) [inline src/EnergyPlus/DataHeatBalance.hh:899]: Few new fields, expected.

File: src/EnergyPlus/InternalHeatGains.cc
Comment (Myoldmopar) [reply src/EnergyPlus/InternalHeatGains.cc:1016]: Fine with deferring until the follow-up issue.

File: testfiles/CMakeLists.txt
Comment (Myoldmopar) [inline testfiles/CMakeLists.txt:30]: New example file :+1:

File: tst/EnergyPlus/unit/ThermalComfort.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/ThermalComfort.unit.cc:1054]: Unit test, excellent.

**Review by Myoldmopar — APPROVED** [review]
This all looks really good.  I am testing locally with updated develop pulled in, but I expect this to merge in a bit.

**Comment by Myoldmopar** [comment]
All good locally, thanks @xuanluo113

**Pull request merged**
